### DISCUSSION TITLE
How to free bufp after call qpdf_oh_get_stream_data in c# p/invoke

### DISCUSSION THREAD

**Discussion by mslichao** [qpdf/qpdf#1278] [Q&A] [answered]
Hi,
I used QPdf C API, and I saw the comments of `qpdf_oh_get_stream_data` said `Get stream data as a buffer. The buffer is allocated with malloc and must be freed by the caller.`
I am using C# p/invoke to call this API, how can I free this memory malloced by native?

**Comment by m-holger** [comment]
Good question. I have no idea how to this from C#. In C, you would call `free(*bufp)`, where `*bufp` is the object pointed to by `bufp` and `free` is defined in header `<stdlib.h>`. I don't know whether that helps. 

@jberkenbilt  , would it make sense to add a `free` function to the C-API to help users calling the API from some third language?

**Reply by jberkenbilt** [reply]
I don't think it would make sense to add a free function to our API. Users should be able to directly call free from the C standard library using the same mechanism that they would use to call our API. We don't need to specifically wrap it. If someone with more experience in using C APIs can contradict me, I'd be open, but I'm nearly certain about this, and this question has never come up before in all the years of the C API.

**Reply by m-holger** [reply]
Generally, to use a shared library from a third language you don't need access to the C standard library nor knowledge of C. Personally, I don't know how to call malloc from either Python or Rust even though I have used a variety of shared libraries exposing a C-API from those languages.

**Reply by jberkenbilt** [reply, marked as answer]
Well, if it makes it easier, I suppose there's no harm in adding a one-liner like qpdf_free that's just a wrapper around free. Or maybe qpdf_free_buffer that takes unsigned char**, frees *p, and sets it to null. It's a harmless addition.

**Reply by mslichao** [reply]
Thanks for your answer.
I create a PR to add this API, can you help to review this PR #1287 ?
In this PR, I manually add one line in the qpdf.testcov, but I am not sure if it is correct.

**Comment by jberkenbilt** [comment]
We'll want this in the ChangeLog/Release notes. @m-holger Usually I have followed up afterwards to add that and credited the contributor in ChangeLog (but not in release notes, which don't need to be cluttered with that).

**Reply by m-holger** [reply]
I was planning to do the same.
